File: dotNETProject-RuleEngine/ConditionList.cs
using System;
using System.Collections.Generic;

namespace Grynsoft.RuleEngine
{
    // A list of conditions evaluated with AND or OR logic.
    // Supports target capture: conditions implementing IConditionTargetProvider can
    // capture Owner/Receiver values that other conditions in the same list can read.
    [Serializable]
    public class ConditionList
    {
        // If true, conditions are OR'd (any true = pass). Default is AND (all must be true).
        public bool UseOrMode { get; set; }

        // The list of conditions.
        public List<ICondition> Conditions { get; set; } = new List<ICondition>();

        // Per-list captures, scoped only to this list.
        [NonSerialized] public object? CapturedOwner;
        [NonSerialized] public object? CapturedReceiver;

        // Initialize all conditions with a master target.
        public void Initialize(object masterTarget)
        {
            foreach (var c in Conditions)
            {
                c?.Initialize(masterTarget);
                if (c is ConditionBase cb) cb.ParentList = this;
            }
        }

        // True if all (or any, in OR mode) conditions are met.
        public bool AreMet => Evaluate();

        // Evaluate all conditions with AND or OR logic.
        public bool Evaluate()
        {
            if (Conditions.Count == 0) return false;
            CapturedOwner = null;
            CapturedReceiver = null;

            if (UseOrMode)
            {
                foreach (var c in Conditions)
                {
                    if (c?.Evaluate() == true)
                    {
                        Capture(c);
                        return true;
                    }
                }
                return false;
            }

            // Two-pass: target providers first (so their captures are visible), then everything else.
            foreach (var c in Conditions)
                if (IsProvider(c) && !EvalAndCapture(c)) return false;

            foreach (var c in Conditions)
                if (!IsProvider(c) && !EvalAndCapture(c)) return false;

            return true;
        }

        // Get a human-readable description of all conditions joined by AND/OR.
        public string GetDescription()
        {
            if (Conditions.Count == 0) return "No conditions set";
            if (Conditions.Count == 1) return Conditions[0].GetDescription();

            string joiner = UseOrMode ? " OR " : " AND ";
            var descriptions = new List<string>();
            foreach (var condition in Conditions)
                descriptions.Add(condition.GetDescription());
            return string.Join(joiner, descriptions);
        }

        private bool EvalAndCapture(ICondition c)
        {
            if (c?.Evaluate() != true) return false;
            Capture(c);
            return true;
        }

        private static bool IsProvider(ICondition c) => c is IConditionTargetProvider p && p.ProvidesTarget;

        private void Capture(ICondition c)
        {
            if (c is not IConditionTargetProvider p) return;
            if (p.CapturedOwner != null) CapturedOwner = p.CapturedOwner;
            if (p.CapturedReceiver != null) CapturedReceiver = p.CapturedReceiver;
        }
    }
}

File: Web-SystemsShowcase/core/src/conditions/ConditionList.ts
// Declarative condition evaluation. A ConditionList holds conditions that are
// checked against a shared context object - used to drive job milestone
// progress and failure.

export interface ICondition {
  myObj: unknown;
  invertCondition: boolean;
  initialize(target: unknown): void;
  evaluate(): boolean;
  getDescription(): string;
}

// Conditions implementing this capture Owner/Receiver values that other
// conditions in the same list can read.
export interface IConditionTargetProvider {
  readonly providesTarget: boolean;
  readonly capturedOwner: unknown;
  readonly capturedReceiver: unknown;
}

export abstract class ConditionBase implements ICondition {
  // Target override - when null, initialize() falls back to the master target.
  target: unknown = null;
  invertCondition = false;
  myObj: unknown = null;

  // Back-reference to the owning list, for reading captures.
  parentList: ConditionList | null = null;

  initialize(masterTarget: unknown): void {
    this.myObj = this.target ?? masterTarget;
  }

  // Subclasses may read parentList captures here.
  protected resolveDynamicTarget(): unknown {
    return this.myObj;
  }

  abstract evaluate(): boolean;
  abstract getDescription(): string;

  protected applyInversion(result: boolean): boolean {
    return this.invertCondition ? !result : result;
  }
}

// Convenience condition wrapping a plain predicate.
export class FuncCondition extends ConditionBase {
  constructor(
    private readonly predicate: (target: unknown) => boolean,
    private readonly description = "custom condition",
  ) {
    super();
  }

  evaluate(): boolean {
    return this.applyInversion(this.predicate(this.resolveDynamicTarget()));
  }

  getDescription(): string {
    return this.invertCondition ? `NOT (${this.description})` : this.description;
  }
}

// A list of conditions evaluated with AND or OR logic.
// Supports target capture: provider conditions evaluate first so their
// captures are visible to the rest of the list.
export class ConditionList {
  // If true, conditions are OR'd (any true = pass). Default is AND.
  useOrMode = false;
  conditions: ICondition[] = [];

  // Per-list captures, scoped only to this list.
  capturedOwner: unknown = null;
  capturedReceiver: unknown = null;

  initialize(masterTarget: unknown): void {
    for (const c of this.conditions) {
      c?.initialize(masterTarget);
      if (c instanceof ConditionBase) c.parentList = this;
    }
  }

  get areMet(): boolean {
    return this.evaluate();
  }

  evaluate(): boolean {
    if (this.conditions.length === 0) return false;
    this.capturedOwner = null;
    this.capturedReceiver = null;

    if (this.useOrMode) {
      for (const c of this.conditions) {
        if (c?.evaluate() === true) {
          this.capture(c);
          return true;
        }
      }
      return false;
    }

    // Two-pass: target providers first (so their captures are visible), then everything else.
    for (const c of this.conditions)
      if (ConditionList.isProvider(c) && !this.evalAndCapture(c)) return false;

    for (const c of this.conditions)
      if (!ConditionList.isProvider(c) && !this.evalAndCapture(c)) return false;

    return true;
  }

  getDescription(): string {
    if (this.conditions.length === 0) return "No conditions set";
    if (this.conditions.length === 1) return this.conditions[0].getDescription();
    const joiner = this.useOrMode ? " OR " : " AND ";
    return this.conditions.map((c) => c.getDescription()).join(joiner);
  }

  private evalAndCapture(c: ICondition): boolean {
    if (c?.evaluate() !== true) return false;
    this.capture(c);
    return true;
  }

  private static isProvider(c: ICondition): boolean {
    return (c as Partial<IConditionTargetProvider>).providesTarget === true;
  }

  private capture(c: ICondition): void {
    const p = c as Partial<IConditionTargetProvider>;
    if (p.capturedOwner != null) this.capturedOwner = p.capturedOwner;
    if (p.capturedReceiver != null) this.capturedReceiver = p.capturedReceiver;
  }
}
